authorizer-app-backend/src/main/java/org/radarbase/authorizer/service/OuraAuthorizationService.kt

```diff
@@ -3,20 +3,54 @@ package org.radarbase.authorizer.service
 import com.fasterxml.jackson.databind.ObjectMapper
 import jakarta.ws.rs.core.Context
 import jakarta.ws.rs.core.UriBuilder
-import okhttp3.Credentials
 import okhttp3.FormBody
 import okhttp3.OkHttpClient
 import okhttp3.Request
+import org.radarbase.authorizer.api.OuraAuthUserId
+import org.radarbase.authorizer.api.RequestTokenPayload
+import org.radarbase.authorizer.api.RestOauth2AccessToken
 import org.radarbase.authorizer.config.AuthorizerConfig
 import org.radarbase.authorizer.doa.entity.RestSourceUser
+import org.radarbase.jersey.exception.HttpBadGatewayException
 import org.radarbase.jersey.util.request
+import org.radarbase.jersey.util.requestJson
 
 class OuraAuthorizationService(
     @Context private val clients: RestSourceClientService,
     @Context private val httpClient: OkHttpClient,
     @Context private val objectMapper: ObjectMapper,
     @Context private val config: AuthorizerConfig,
 ) : OAuth2RestSourceAuthorizationService(clients, httpClient, objectMapper, config) {
+    private val tokenReader = objectMapper.readerFor(RestOauth2AccessToken::class.java)
+    private val oauthUserReader = objectMapper.readerFor(OuraAuthUserId::class.java)
+
+    override fun requestAccessToken(payload: RequestTokenPayload, sourceType: String): RestOauth2AccessToken {
+        val authorizationConfig = clients.forSourceType(sourceType)
+        val clientId = checkNotNull(authorizationConfig.clientId)
+
+        val form = FormBody.Builder().apply {
+            payload.code?.let { add("code", it) }
+            add("grant_type", "authorization_code")
+            add("client_id", clientId)
+            add("redirect_uri", config.service.callbackUrl.toString())
+        }.build()
+        val accessToken: RestOauth2AccessToken = httpClient.requestJson(post(form, sourceType), tokenReader)
+        if (accessToken.accessToken == null) {
+            logger.error("Failed to get access token for user {}", clientId)
+            throw HttpBadGatewayException("Service $sourceType did not provide a result")
+        }
+        val ouraUserUri = UriBuilder.fromUri(Oura_USER_ID_ENDPOINT).queryParam("access_token", accessToken.accessToken).build().toString()
```

Comment:
Could we move this to a separate function and have the `accessToken` as the parameter?